cache/test_cache_control.py

```diff
@@ -806,9 +809,53 @@ class HttpChainCacheActionCached(TestCacheControl, SingleTest):
             cookie "foo_items_in_cart" == "*" -> $$cache = 0;
             cookie "comment_author_*" == "*" -> $$cache = 1;
             cookie "wordpress_logged_in*" == "*" -> $$cache = 0;
-            -> vh1;
+            -> default;
         }
         '''
     request_headers = {'Cookie': 'comment_author_name=john'}
     response_headers = {}
     should_be_cached = True
+
+class CacheLocationBase(TestCacheControl, SingleTest, base=True):
+    tempesta_template = {
+        "config":
+        """
+        server ${server_ip}:8000;
+        vhost default {
+            proxy_pass default;
+
+            location prefix "/cached" {
+                proxy_pass default;
+                cache_fulfill * *;
+            }
+
+            location prefix "/bypassed" {
+                proxy_pass default;
+                cache_bypass * *;
+            }
+            location prefix "/nonidempotent" {
+                proxy_pass default;
+                cache_fulfill * *;
+                nonidempotent GET * *;
```

Comment:
What is our expected behavior here? From my understanding, if a request is explicitly marked as a non-idempotent one, that literally means that it cannot be cached, since it might modify some internal backend state and following responses would differ, right? As for now, it gets cached.

Reply:
Quite interesting question. Technically, we can cache responses for non-idempotent requests to service the responses later for idempotent requests: e.g. cache a response for POST and later use the cached response for GETs. However, in your example GETs are non-idempotent.

Since a non-idempotent request is considered to change a server state, we must not service it from the cache. I.e. `tfw_cache_msg_cacheable()` should also check the request with `tfw_http_req_is_nip()` (both the functions are static for their C files, so probably the second one should be moved to a header).

With this test I'd test that GET aren't serviced from the cache, but for example HEAD do use the cache.

Reply:
Now we have https://github.com/tempesta-tech/tempesta/issues/1663 for this